File: windows/tests/Elevate.Core.Tests/Managed/ManagedTenantResolverTests.cs
using Elevate.Core.Managed;
using Elevate.Core.Tests.Support;
using FluentAssertions;

namespace Elevate.Core.Tests.Managed;

public class ManagedTenantResolverTests
{
    private static string Issuer(string tenantId)
        => $$"""{"issuer":"https://login.microsoftonline.com/{{tenantId}}/v2.0"}""";

    [Fact]
    public async Task GuidPassesThroughWithoutARequest()
    {
        var http = new StubHttpClient();
        var resolver = new ManagedTenantResolver(http);

        var result = await resolver.ResolveAsync([" 11111111-2222-3333-4444-AAAAAAAAAAAA "]);

        result.Ids.Should().Equal(new Dictionary<string, string>
        {
            [" 11111111-2222-3333-4444-AAAAAAAAAAAA "] = "11111111-2222-3333-4444-aaaaaaaaaaaa",
        });
        result.Unresolved.Should().BeEmpty();
        http.Requests.Should().BeEmpty();
    }

    [Fact]
    public async Task ADomainIsRequestedOnlyOncePerResolver()
    {
        var http = new StubHttpClient();
        http.On("GET", "fabrikam.com/v2.0/.well-known/openid-configuration", Issuer("11111111-2222-3333-4444-555555555555"));
        var resolver = new ManagedTenantResolver(http);

        var first = await resolver.ResolveAsync(["fabrikam.com"]);
        var second = await resolver.ResolveAsync(["fabrikam.com"]);

        first.Ids["fabrikam.com"].Should().Be("11111111-2222-3333-4444-555555555555");
        second.Ids["fabrikam.com"].Should().Be("11111111-2222-3333-4444-555555555555");
        http.RequestsMatching("openid-configuration").Should().ContainSingle();
    }

    [Fact]
    public async Task RepeatedEntriesInOneCallAreRequestedOnce()
    {
        var http = new StubHttpClient();
        http.On("GET", "openid-configuration", Issuer("11111111-2222-3333-4444-555555555555"));
        var resolver = new ManagedTenantResolver(http);

        var result = await resolver.ResolveAsync(["fabrikam.com", "fabrikam.com"]);

        result.Ids.Should().ContainSingle();
        http.RequestsMatching("openid-configuration").Should().ContainSingle();
    }

    /// <summary>
    /// The policy may say <c>contoso.com</c> where a published profile says <c>Contoso.com</c>.
    /// Both must resolve, and the second spelling must not cost a second request.
    /// </summary>
    [Fact]
    public async Task TwoSpellingsOfADomainResolveWithOneRequest()
    {
        var http = new StubHttpClient();
        http.On("GET", "fabrikam.com/v2.0/.well-known/openid-configuration", Issuer("11111111-2222-3333-4444-555555555555"));
        var resolver = new ManagedTenantResolver(http);

        var result = await resolver.ResolveAsync(["fabrikam.com", "Fabrikam.COM"]);

        result.Ids["fabrikam.com"].Should().Be("11111111-2222-3333-4444-555555555555");
        result.Ids["Fabrikam.COM"].Should().Be("11111111-2222-3333-4444-555555555555");
        result.Unresolved.Should().BeEmpty();
        http.RequestsMatching("openid-configuration").Should().ContainSingle();
    }

    [Fact]
    public async Task UnknownDomainLandsInUnresolved()
    {
        var http = new StubHttpClient();
        http.On("GET", "openid-configuration", """{"error":"invalid_tenant"}""", status: 404);
        var resolver = new ManagedTenantResolver(http);

        var result = await resolver.ResolveAsync(["nope.example"]);

        result.Ids.Should().BeEmpty();
        result.Unresolved.Should().Equal("nope.example");
    }

    [Fact]
    public async Task AFailureIsNotCachedSoTheNextCallRetries()
    {
        var http = new StubHttpClient();
        http.On("GET", "openid-configuration", string.Empty, status: 500);
        var resolver = new ManagedTenantResolver(http);

        await resolver.ResolveAsync(["fabrikam.com"]);
        http.On("GET", "openid-configuration", Issuer("11111111-2222-3333-4444-555555555555"));
        var second = await resolver.ResolveAsync(["fabrikam.com"]);

        second.Ids["fabrikam.com"].Should().Be("11111111-2222-3333-4444-555555555555");
        http.RequestsMatching("openid-configuration").Should().HaveCount(2);
    }

    [Fact]
    public async Task MixedInputResolvesWhatItCanAndKeepsUnresolvedOrder()
    {
        var http = new StubHttpClient();
        http.On("GET", "fabrikam.com/v2.0", Issuer("11111111-2222-3333-4444-555555555555"));
        http.On("GET", "nope.example/v2.0", string.Empty, status: 404);
        http.On("GET", "also-bad.example/v2.0", string.Empty, status: 404);
        var resolver = new ManagedTenantResolver(http);

        var result = await resolver.ResolveAsync(
        [
            "nope.example",
            "22222222-3333-4444-5555-666666666666",
            "fabrikam.com",
            "also-bad.example",
        ]);

        result.Ids.Keys.Should().BeEquivalentTo("22222222-3333-4444-5555-666666666666", "fabrikam.com");
        result.Ids["fabrikam.com"].Should().Be("11111111-2222-3333-4444-555555555555");
        result.Ids["22222222-3333-4444-5555-666666666666"].Should().Be("22222222-3333-4444-5555-666666666666");
        result.Unresolved.Should().Equal("nope.example", "also-bad.example");
    }

    [Fact]
    public async Task AnUnparseableBodyIsUnresolved()
    {
        var http = new StubHttpClient();
        http.On("GET", "openid-configuration", "not json at all");
        var resolver = new ManagedTenantResolver(http);

        (await resolver.ResolveAsync(["broken.example"])).Unresolved.Should().Equal("broken.example");
    }

    [Fact]
    public async Task AnIssuerWithoutATenantIdIsUnresolved()
    {
        var http = new StubHttpClient();
        http.On("GET", "openid-configuration", """{"issuer":"https://login.microsoftonline.com/common/v2.0"}""");
        var resolver = new ManagedTenantResolver(http);

        (await resolver.ResolveAsync(["common.example"])).Unresolved.Should().Equal("common.example");
    }
}

File: macos/Tests/ElevateCoreTests/ManagedTenantResolverTests.swift
import Testing
import Foundation
@testable import ElevateCore

@Suite struct ManagedTenantResolverTests {
    private func issuer(_ tenantId: String) -> Data {
        Data(#"{"issuer":"https://login.microsoftonline.com/\#(tenantId)/v2.0"}"#.utf8)
    }

    @Test func guidPassesThroughWithoutARequest() async {
        let http = StubHTTPClient()
        let resolver = ManagedTenantResolver(http: http)

        let result = await resolver.resolve([" 11111111-2222-3333-4444-AAAAAAAAAAAA "])

        #expect(result.ids == [" 11111111-2222-3333-4444-AAAAAAAAAAAA ": "11111111-2222-3333-4444-aaaaaaaaaaaa"])
        #expect(result.unresolved.isEmpty)
        #expect(await http.requests.isEmpty)
    }

    @Test func aDomainIsRequestedOnlyOncePerResolver() async {
        let http = StubHTTPClient()
        await http.on("GET", "fabrikam.com/v2.0/.well-known/openid-configuration",
                      body: issuer("11111111-2222-3333-4444-555555555555"))
        let resolver = ManagedTenantResolver(http: http)

        let first = await resolver.resolve(["fabrikam.com"])
        let second = await resolver.resolve(["fabrikam.com"])

        #expect(first.ids["fabrikam.com"] == "11111111-2222-3333-4444-555555555555")
        #expect(second.ids["fabrikam.com"] == "11111111-2222-3333-4444-555555555555")
        #expect(await http.requests(matching: "openid-configuration").count == 1)
    }

    @Test func repeatedEntriesInOneCallAreRequestedOnce() async {
        let http = StubHTTPClient()
        await http.on("GET", "openid-configuration", body: issuer("11111111-2222-3333-4444-555555555555"))
        let resolver = ManagedTenantResolver(http: http)

        let result = await resolver.resolve(["fabrikam.com", "fabrikam.com"])

        #expect(result.ids.count == 1)
        #expect(await http.requests(matching: "openid-configuration").count == 1)
    }

    /// The policy may say `contoso.com` where a published profile says `Contoso.com`. Both must
    /// resolve, and the second spelling must not cost a second request.
    @Test func twoSpellingsOfADomainResolveWithOneRequest() async {
        let http = StubHTTPClient()
        await http.on("GET", "fabrikam.com/v2.0/.well-known/openid-configuration",
                      body: issuer("11111111-2222-3333-4444-555555555555"))
        let resolver = ManagedTenantResolver(http: http)

        let result = await resolver.resolve(["fabrikam.com", "Fabrikam.COM"])

        #expect(result.ids["fabrikam.com"] == "11111111-2222-3333-4444-555555555555")
        #expect(result.ids["Fabrikam.COM"] == "11111111-2222-3333-4444-555555555555")
        #expect(result.unresolved.isEmpty)
        #expect(await http.requests(matching: "openid-configuration").count == 1)
    }

    @Test func unknownDomainLandsInUnresolved() async {
        let http = StubHTTPClient()
        await http.on("GET", "openid-configuration", status: 404, body: Data(#"{"error":"invalid_tenant"}"#.utf8))
        let resolver = ManagedTenantResolver(http: http)

        let result = await resolver.resolve(["nope.example"])

        #expect(result.ids.isEmpty)
        #expect(result.unresolved == ["nope.example"])
    }

    @Test func aFailureIsNotCachedSoTheNextCallRetries() async {
        let http = StubHTTPClient()
        await http.on("GET", "openid-configuration", status: 500, body: Data())
        let resolver = ManagedTenantResolver(http: http)

        _ = await resolver.resolve(["fabrikam.com"])
        await http.on("GET", "openid-configuration", body: issuer("11111111-2222-3333-4444-555555555555"))
        let second = await resolver.resolve(["fabrikam.com"])

        #expect(second.ids["fabrikam.com"] == "11111111-2222-3333-4444-555555555555")
        #expect(await http.requests(matching: "openid-configuration").count == 2)
    }

    @Test func mixedInputResolvesWhatItCanAndKeepsUnresolvedOrder() async {
        let http = StubHTTPClient()
        await http.on("GET", "fabrikam.com/v2.0", body: issuer("11111111-2222-3333-4444-555555555555"))
        await http.on("GET", "nope.example/v2.0", status: 404, body: Data())
        await http.on("GET", "also-bad.example/v2.0", status: 404, body: Data())
        let resolver = ManagedTenantResolver(http: http)

        let result = await resolver.resolve([
            "nope.example",
            "22222222-3333-4444-5555-666666666666",
            "fabrikam.com",
            "also-bad.example",
        ])

        #expect(Set(result.ids.keys) == ["22222222-3333-4444-5555-666666666666", "fabrikam.com"])
        #expect(result.ids["fabrikam.com"] == "11111111-2222-3333-4444-555555555555")
        #expect(result.ids["22222222-3333-4444-5555-666666666666"] == "22222222-3333-4444-5555-666666666666")
        #expect(result.unresolved == ["nope.example", "also-bad.example"])
    }

    @Test func anUnparseableBodyIsUnresolved() async {
        let http = StubHTTPClient()
        await http.on("GET", "openid-configuration", body: Data("not json at all".utf8))
        let resolver = ManagedTenantResolver(http: http)

        #expect(await resolver.resolve(["broken.example"]).unresolved == ["broken.example"])
    }

    @Test func anIssuerWithoutATenantIdIsUnresolved() async {
        let http = StubHTTPClient()
        await http.on("GET", "openid-configuration",
                      body: Data(#"{"issuer":"https://login.microsoftonline.com/common/v2.0"}"#.utf8))
        let resolver = ManagedTenantResolver(http: http)

        #expect(await resolver.resolve(["common.example"]).unresolved == ["common.example"])
    }
}
